azurelinuxagent/ga/cgroupapi.py

```diff
@@ -37,10 +37,32 @@
 from azurelinuxagent.common.version import get_distro
 
 CGROUPS_FILE_SYSTEM_ROOT = '/sys/fs/cgroup'
-CGROUP_CONTROLLERS = ["cpu", "memory"]
 EXTENSION_SLICE_PREFIX = "azure-vmextensions"
 
 
+def get_cgroup_api():
+    """
+    Determines which version of Cgroups should be used for resource enforcement and monitoring by the Agent are returns
+    the corresponding Api. If the required controllers are not mounted in v1 or v2, return None.
+    """
+    v1 = SystemdCgroupsApiv1()
+    v2 = SystemdCgroupsApiv2()
+
+    log_cgroup_info("Controllers mounted in v1: {0}. Controllers mounted in v2: {1}".format(v1.get_mounted_controllers(), v2.get_mounted_controllers()))
+
+    # It is possible for different controllers to be simultaneously mounted under v1 and v2. If any are mounted under
+    # v1, use v1.
+    if v1.is_cpu_or_memory_mounted():
```

Comment:
is this supposed to be and? we use v1 if both cpu and memory mounted?

I just thought one scenario we use v1 for cpu, and cx enabled memory in v2 for their application. Then we start using v1 for memory then that will create discprencies

Reply:
Seems like we simplify this asking for the file system type

```If you wonder how to detect which of these three modes is currently used, use statfs() on /sys/fs/cgroup/. If it reports CGROUP2_SUPER_MAGIC in its .f_type field, then you are in unified mode. If it reports TMPFS_MAGIC then you are either in legacy or hybrid mode. To distinguish these two cases, run statfs() again on /sys/fs/cgroup/unified/. If that succeeds and reports CGROUP2_SUPER_MAGIC you are in hybrid mode, otherwise not. From a shell, you can check the Type in stat -f /sys/fs/cgroup and stat -f /sys/fs/cgroup/unified.```

Then if unified use v2, else v1.

Results from ubuntu 20 (hybrid) and 22 (unified)

```
root@nam-u20:/home/nam# stat -f --format=%T /sys/fs/cgroup
tmpfs

root@nam-ubuntu22:/home/nam# stat -f --format=%T /sys/fs/cgroup
cgroup2fs
```